### PR TITLE
Feat: add SMTP support 

### PR DESCRIPTION
**Pull request by Tailp**
Hi @monperrus @danglotb :) , final feature for web interface. After this we can amp files in email to users. 

After this PR, I'll make 2 last PRs updating docker image of Dspot and webinterface involving bugfixes for those parts.

### CODE REVIEW

File: dspot/src/main/java/eu/stamp_project/utils/collector/CollectorConfig.java [hunk 1]
```diff
@@ -21,6 +21,8 @@ public static CollectorConfig getInstance() {
 
     /* Variables and getters/setters */
     private static DspotInformationCollector collector = new NullCollector();
+    private static String gmailUserName;
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/utils/collector/CollectorConfig.java]: we don't want to depend on GMail, it should be `smtpUserName`

File: dspot/src/main/java/eu/stamp_project/utils/smtp/EmailSender.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            // Set From: header field of the header.
+            message.setFrom(new InternetAddress(fromUser));
+
+            // Set To: header field of the header.
+            message.setRecipients(Message.RecipientType.TO,InternetAddress.parse(toUsers));
+
+            // Set Subject: header field
+            message.setSubject(subject);
+
+            message.setText(messageText);
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/utils/smtp/EmailSender.java]: where's the email template?

File: dspot/src/main/java/eu/stamp_project/utils/collector/CollectorConfig.java (hunk 1 above)
Comment (Tailp) [reply dspot/src/main/java/eu/stamp_project/utils/collector/CollectorConfig.java]: Now gmail has become the more general smtp :) .

File: dspot/src/main/java/eu/stamp_project/utils/smtp/EmailSender.java (hunk 2 above)
Comment (Tailp) [reply dspot/src/main/java/eu/stamp_project/utils/smtp/EmailSender.java]: you mean how the messageText is constructed ? . There is a method in `MongodbCollector.java` called `constructMessageWithFileContents` which handle that at the end process. Currently it only appends String .

**Comment by Tailp** [comment]
Resolved :) . Please have a look again.

File: dspot/src/main/java/eu/stamp_project/utils/collector/CollectorConfig.java (hunk 1 above)
Comment (monperrus) [reply dspot/src/main/java/eu/stamp_project/utils/collector/CollectorConfig.java]: all variables called `gmail*` should be renamed `smtp*`

Comment (Tailp) [reply dspot/src/main/java/eu/stamp_project/utils/collector/CollectorConfig.java]: Done :) . Those are redundant lines . I forgot to remove them.

**Comment by Tailp** [comment]
@monperrus ready for comment :) .

**Comment by monperrus** [comment]
The responsibility of `MongodbCollector` is to push data in MongoDB, not to send email.s

According to the [single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle) this feature should be in a separate collector `EmailSenderCollector`.

**Comment by Tailp** [comment]
@monperrus If we create a EmailSenderCollector, I think we would end up with replicating the code in mongodb or using 2 different collector type during the same execution. I think It's better that we remove the emailing part in MongodbCollector itself and put it as a step in main instead, which achieve the purpose of 1 purpose per module :) .

**Comment by monperrus** [comment]
We can keep it this way for now. Could you add a test case checking that the feature works as expected?

**Comment by Tailp** [comment]
@monperrus hi, can we somehow hide the password and username of the email , since I don't know how to test it actually sending an email. We already have a restful mode test previously just that the email information is not set in that test due to security reason :).

**Comment by monperrus** [comment]
This is an interesting testability issue. You need to use a fake server or to mock.

See for example:
https://stackoverflow.com/questions/1027595/running-junit-email-tests-without-emails-actually-going-out

**Comment by Tailp** [comment]
Hi @monperrus :), I made the EmailSender class a singleton then check the result of mail sending after the execution of restful mode, to verify that it actually tried to sent an email during the run. This work more or less like a mock.

**Comment by monperrus** [comment]
this is a good start, you test the error mode. now, we should test the happy path (sending an email is working)

**Comment by Tailp** [comment]
@monperrus hi :) , I happen to stumble upon this [fake smtp server](https://www.smtpbucket.com/) now the test will be sending email there instead. We can check that it actually was sended by typing in the receiver `abc@mail.com` on the mentioned website.

**Comment by monperrus** [comment]
something in that direction, but with no external service.

Something like https://github.com/voodoodyne/subethasmtp ([doc](https://github.com/voodoodyne/subethasmtp/blob/master/Wiser.md))

**Comment by Tailp** [comment]
@monperrus hi :) , I understand. I removed the previous code and use GreenMail instead and set up a tempo smtp server for testing, which is better. Ready for reviews .

**Comment by monperrus** [comment]
Thanks a lot, LGTM. @danglotb merge?

**Comment by Tailp** [comment]
@monperrus  @danglotb  Hi :) , It would be great if we make a [release](https://github.com/STAMP-project/dspot/releases/) with these new feature so that I can smoothly download the jar each time I build the image instead of building it and add it locally .

**Pull request merged**